crates/sui-core/src/authority/authority_store_tables.rs

```diff
@@ -137,14 +137,16 @@ impl AuthorityPerpetualTables {
         iter.reverse().next().and_then(|(_, o)| self.object(o).ok())
     }
 
-    pub fn object(&self, store_object: StoreObject) -> Result<Object, SuiError> {
+    pub fn object(&self, store_object: StoreObjectWrapper) -> Result<Object, SuiError> {
+        let store_object = store_object.migrate().into_inner();
         let indirect_object = match store_object.data {
```

Comment:
I don't think this code will have any negative performance impact. The overwhelmingly common case will be that objects will already be in the latest version, except for a brief period after a new version is released, in which case some form of migration will be necessary no matter what. A compaction filter could work too, although I am (perhaps misguidedly) trying to keep us from becoming too tied to rocksdb.

Reply:
compaction filter approach is quite neat. It won't enable complex migrations either, but should work well for simple migrations. 
We won't even have to keep enum wrapper with old formats in the main code. Just object struct with the version field should work.
Compaction filters operate on binary arrays, so we can utilize partial deserialisation to check the version.
In this way we could keep old formats visible only for compaction filter and don't pollute the rest of the codebase

Reply:
@phoenix-o BCS encoding of enum already stores the version as the very first byte in the serialized payload. Check this test out from the link I posted above:
```
#[derive(Serialize)]
enum E {
    Variant0(u16),
    Variant1(u8),
    Variant2(String),
}

let v0 = E::Variant0(8000);
let v1 = E::Variant1(255);
let v2 = E::Variant2("e".to_owned());

assert_eq!(to_bytes(&v0)?, vec![0, 0x40, 0x1F]);
assert_eq!(to_bytes(&v1)?, vec![1, 0xFF]);
assert_eq!(to_bytes(&v2)?, vec![2, 1, b'e']);
```
So, encoding this as enum achieves what you are suggesting i.e. partial deserialization by reading just the first byte of the serialized object (that compaction filter passes to the api). I think the enum approach would be consistent with the rest of codebase (we are doing this in bunch of different places to support schema evolution)

Reply:
@sadhansood: yeah, I know and totally agree it'd work, but I was thinking to use version field to get rid of such enum completely. 
I don't see much value of keeping Variant0, Variant1 types in the main code when it's no longer needed (or needed just for the sake of migration).
I buy the consistency argument though, no point to special case objects. I guess I'd just prefer to avoid versioning enums altogether:)